### PR TITLE
Sigma plot

### PR DESCRIPTION
**Pull request by JavierSijen**

### CODE REVIEW

**Comment by JensWehner** [comment]
Want me to comment on the todos?

**Comment by JensWehner** [comment]
Whaat kind of parameters should states take "all" "1,2,3" and "1:300"?

**Comment by JavierSijen** [comment]
> 
> 
> Want me to comment on the todos?

Yes, if you want

**Comment by JavierSijen** [comment]
> 
> 
> Whaat kind of parameters should states take "all" "1,2,3" and "1:300"?

Yeah, that's the plan. I'm using RangeParser atm, which handles "1,2,3" and "1:300". Though currently I'm using zero-based indexing. I'm not sure yet how to handle "all".

**Comment by JavierSijen** [comment]
This is the output. Grid point 0 corresponds to the GW energy after the GW calculation.
`
grid point	omega(0)	sigma(0)	omega(1)	sigma(1)	omega(2)	sigma(2)
 -50	-11.0210	 +0.7754	-11.0211	 +0.7750	 -1.3536	 +0.3383
 -49	-11.0110	 +0.6892	-11.0111	 +0.6894	 -1.3436	 +0.3417
 -48	-11.0010	 +0.7558	-11.0011	 +0.7520	 -1.3336	 +0.2516
...
  -1	-10.5310	 +0.4992	-10.5311	 +0.4991	 -0.8636	 +0.1368
   0	-10.5210	 +0.4957	-10.5211	 +0.4956	 -0.8536	 +0.1338
   1	-10.5110	 +0.4924	-10.5111	 +0.4922	 -0.8436	 +0.1309
...
  48	-10.0410	 +0.3856	-10.0411	 +0.3856	 -0.3736	 +0.0557
  49	-10.0310	 +0.3840	-10.0311	 +0.3840	 -0.3636	 +0.0546
  50	-10.0210	 +0.3824	-10.0211	 +0.3824	 -0.3536	 +0.0536
`

**Comment by JensWehner** [comment]
you do not need the gridpoint line in the output.

File: src/libxtp/gwbse/gw.cc [hunk 1]
```diff
@@ -300,5 +303,44 @@ void GW::CalculateHQP() {
...
+  const Index steps = _opt.sigma_plot_steps;
+  const double spacing = _opt.sigma_plot_spacing;
+  const Index qptotal = _opt.qpmax - _opt.qpmin + 1;
+  XTP_LOG(Log::info, _log) << TimeStamp() << " Plotting Sigma diagonals "
+                           << std::flush;
+  tools::RangeParser rp;
+  rp.Parse(_opt.sigma_plot_states);  // TODO: Error handling?
+  // TODO: How to handle "all"?
+  std::ofstream out;
+  out.open("sigma_plot.txt");
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: this should be an option

File: src/libxtp/gwbse/gw.cc [hunk 2]
```diff
@@ -300,5 +303,44 @@ void GW::CalculateHQP() {
...
 
+void GW::PlotSigma(const Eigen::VectorXd& frequencies) const {
+  const Index steps = _opt.sigma_plot_steps;
+  const double spacing = _opt.sigma_plot_spacing;
+  const Index qptotal = _opt.qpmax - _opt.qpmin + 1;
+  XTP_LOG(Log::info, _log) << TimeStamp() << " Plotting Sigma diagonals "
+                           << std::flush;
+  tools::RangeParser rp;
+  rp.Parse(_opt.sigma_plot_states);  // TODO: Error handling?
+  // TODO: How to handle "all"?
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: just check if the string is all and if create a range with all states?

File: src/libxtp/gwbse/gwbse.cc [hunk 3]
```diff
@@ -405,6 +405,25 @@ void GWBSE::Initialize(tools::Property& options) {
...
 
+  _gwopt.sigma_plot_states =
+      options.ifExistsReturnElseReturnDefault<std::string>(
+          key + ".sigma_plot_states", _gwopt.sigma_plot_states);
+  _gwopt.sigma_plot_steps = options.ifExistsReturnElseReturnDefault<Index>(
+      key + ".sigma_plot_steps", _gwopt.sigma_plot_steps);
+  _gwopt.sigma_plot_spacing = options.ifExistsReturnElseReturnDefault<double>(
+      key + ".sigma_plot_spacing", _gwopt.sigma_plot_spacing);
+  // TODO: Option for filename?
+  // TODO: Option to write as binary file?
```
Comment (JensWehner) [inline src/libxtp/gwbse/gwbse.cc]: binary is not needed.

File: src/libxtp/gwbse/gw.cc [hunk 4]
```diff
@@ -300,5 +303,58 @@ void GW::CalculateHQP() {
...
+  const Index steps = _opt.sigma_plot_steps;
+  const double spacing = _opt.sigma_plot_spacing;
+  const Index qptotal = _opt.qpmax - _opt.qpmin + 1;
+  XTP_LOG(Log::info, _log) << TimeStamp() << " Plotting Sigma diagonals "
+                           << std::flush;
+
+  // TODO: Can the following vector be filled more elegantly?
+  std::vector<int> ind;
+  if (_opt.sigma_plot_states == "all") {
+    for (Index gw_level = 0; gw_level < qptotal; gw_level++) {
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: a) use Index instead of int b) start at qpmin c) std::iota

File: src/libxtp/gwbse/gw.cc [hunk 5]
```diff
@@ -300,5 +303,58 @@ void GW::CalculateHQP() {
...
+      ind.push_back(gw_level);
+    }
+  } else {
+    tools::RangeParser rp;
+    rp.Parse(_opt.sigma_plot_states);  // TODO: Error handling?
+    for (Index gw_level : rp) {
+      if (gw_level >= 0 && gw_level < qptotal) {
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: qpmin

Comment (JensWehner) [reply src/libxtp/gwbse/gw.cc]: so this has to be corrected

File: src/libxtp/gwbse/gw.cc [hunk 6]
```diff
@@ -300,5 +303,56 @@ void GW::CalculateHQP() {
...
+               (i == 0 ? "" : "\t") % gw_level;
+  }
+  out << std::endl;
+
+  boost::format numFormat("%+1.4f");
+  Eigen::IOFormat matFormat(Eigen::StreamPrecision, 0, "", "\t");
+  for (Index grid_point = 0; grid_point < steps; grid_point++) {
+    const double offset = (grid_point - ((steps - 1) / 2)) * spacing;
+    Eigen::VectorXd row = Eigen::VectorXd::Zero(2 * count);
+#pragma omp parallel for schedule(dynamic)
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: can you parallelize over gridpoints not levels?

Comment (JavierSijen) [reply src/libxtp/gwbse/gw.cc]: Yeah, I could. But then I'd have to either (1) transpose the output matrix or (2) keep the entire output matrix in memory and only write when it's completely filled. Right now it writes the file row by row and only ever needs one row in memory. Though I suppose having a 10,000 x qptotal matrix in memory isn't that big of a deal. What do you think?

Comment (JensWehner) [reply src/libxtp/gwbse/gw.cc]: actually to be honest I would scratch the "all" option and keep everything in memory. It is very unlikely that someone wants to print out so many states.

**Comment by JensWehner** [comment]
@JavierSijen can this be merged soon?

File: src/libxtp/gwbse/gw.cc [hunk 7]
```diff
@@ -300,5 +303,54 @@ void GW::CalculateHQP() {
...
+  const Index num_states = state_inds.size();
+
+  Eigen::MatrixXd mat = Eigen::MatrixXd::Zero(num_points, 2 * num_states);
+#pragma omp parallel for schedule(dynamic)
+  for (Index grid_point = 0; grid_point < num_points; grid_point++) {
+    const double offset = (grid_point - ((num_points - 1) / 2)) * spacing;
+    for (Index i = 0; i < num_states; i++) {
+      const Index gw_level = state_inds[i];
+      const double omega = frequencies(gw_level) + offset;
+      const double sigma = _sigma->CalcCorrelationDiagElement(gw_level, omega);
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: we have to print the quasiparticle energy out not just sigma_c

**Pull request merged**